### PR TITLE
feat: Tracks saving

### LINKED ISSUE #239: Saving or discarding tracks
## Saving or discarding tracks
- User clicks on `Stop Tracks` button
- User is invited to save track and add details as a New Observation
- User can edit and save track or discard it
- FYI - On a CoMapeo project, an observation refers to a point on the map collected by CoMapeo Mobile. It is based on a geographic location (it has [coordinates](https://docs.mapeo.app/complete-reference-guide/glossary-and-quick-reference-sheets/glossary-of-mapeo-related-terms#gps-georeference-coordinate-formats-location)) and can have associated photos, notes, and other details.

## Saving tracks
Screen Shot 2024-04-08 at 3 02 12 PM

- [x] Display Tracks as an Observation
- [x] Press checkmark icon in the top right to Save Tracks

## Discarding tracks
If the user taps to Exit we show them our warning modal. This action is destructive so we want to double confirm with the user  that this is indeed the action they want to take.

Screen Shot 2024-04-08 at 3 13 16 PM

- [x] Display warning modal
- [x] Display title in bold *Discard Track?* and warning *Your Track will not be saved. This cannot be undone.*
- [x] Create `Discard Tracks` button
- [x] If `Discard Tracks` button clicked, redirect to the map
- [x] Create `Continue editing` button
- [x] If `Continue editing` button clicked, bring back to the Tracks as a New Observation 

## Design
https://www.figma.com/file/1Knqv1NrijUuqMo1pvYyF7/Tracks-and-Audio-(Share)?type=design&node-id=72-1761&mode=design&t=uliZJlzlAUbMNi91-0

### PR DESCRIPTION
**Pull request by CDFN**
This PR adds hooks for backend communication and implements tracks saving.

Closes #239

### CODE REVIEW

File: src/frontend/hooks/server/track.ts
Comment (achou11) [inline src/frontend/hooks/server/track.ts:41]: check is not necessary since `useProject` ensures that project exists

File: src/frontend/hooks/server/track.ts
Comment (achou11) [inline src/frontend/hooks/server/track.ts:16]: check is not necessary since `useProject` ensures that project exists

File: src/frontend/hooks/server/track.ts
Comment (achou11) [inline src/frontend/hooks/server/track.ts:30]: check is not necessary since `useProject` ensures that project exists

File: src/frontend/hooks/server/track.ts
Comment (achou11) [inline src/frontend/hooks/server/track.ts:52]: check is not necessary since `useProject` ensures that project exists

File: src/frontend/screens/MapScreen/track/saveTrack/DiscardTrackModal.tsx
Comment (achou11) [inline src/frontend/screens/MapScreen/track/saveTrack/DiscardTrackModal.tsx:1]: thinking this should reuse the `BottomSheetModal` component we have implemented in `sharedComponents`. It has properties and behavior that align with design decisions we've agreed on (such as dismissing only by button). the component implementation is not perfect but it works well enough and using it will make it easier to maintain for us later on.

Can elaborate on usage elsewhere if needed, but basically the exports of `sharedComponents/BottomSheetModal/index.tsx` should have what you need to implement this. General usage will look like this:

```tsx
import {useBottomSheetModal, BottomSheetModal, BottomSheetContent} from '...' // wherever sharedComponents/BottomSheetModal lives

function ParentComponent() {
  const {sheetRef, isOpen, openSheet} = useBottomSheetModal({
    openOnMount: false,
  });

  return (
    <>
      { /* Whatever other markup you need for your component */}
      <View>{...}</View>
      <BottomSheetModal isOpen={isOpen} ref={sheetRef}>
        <BottomSheetContent
          titleStyle={}
          descriptionStyle={}
          title={}
          description={}
          buttonConfigs={}
          icon={}
        />
      </BottomSheetModal>
    </>
  )
}
```

**Review by achou11 — CHANGES_REQUESTED** [review]
Mostly a UX review but some feedback about implementation

- fix: indentation of text area should be aligned with category row displayed above

  image

- fix: behavior of discard track bottom sheet modal is not aligned with our design practices around bottom sheet modals. would suggest using the shared component we have to make it easier

File: src/backend/package.json
Comment (bohdanprog) [reply src/backend/package.json:17]: fixed

File: src/frontend/screens/MapScreen/track/saveTrack/DiscardTrackModal.tsx
Comment (bohdanprog) [reply src/frontend/screens/MapScreen/track/saveTrack/DiscardTrackModal.tsx:1]: Hey, I have updated our solution based on your recommendations. Let me know if I need to change something else :)

File: src/frontend/screens/MapScreen/track/SaveTrackScreen.tsx
Comment (ErikSin) [inline src/frontend/screens/MapScreen/track/SaveTrackScreen.tsx:37]: We want to persist the description in case the user closes their phone. instead of using `setDescription` use                 [updateTags('notes', newVal);](https://github.com/digidem/comapeo-mobile/blob/dd9101b1765e5d3398b0ab6f5a231a0ba7f049f9/src/frontend/screens/ObservationEdit/DescriptionField.tsx#L26) and instead of description use `[const notes = usePersistedDraftObservation(store => store.value?.tags.notes)](https://github.com/digidem/comapeo-mobile/blob/dd9101b1765e5d3398b0ab6f5a231a0ba7f049f9/src/frontend/screens/ObservationEdit/DescriptionField.tsx#L18C3-L18C80)`

**Review by ErikSin — APPROVED** [review]
Besides my 2 comments, and andrew's comments, looks good to me.

I have noticed some architecture changes related to the location provider that I think we need to revert, but I will create an issue for that.

**Comment by ErikSin** [comment]
If the user has stopped "tracking" but is in the middle of saving it (aka they are on `SaveTrackScreen`) and they close the app, we want to make sure it opens again on the saveTrackScreen, and that the tracks are persisted in state. In observations acheive this by [saving observations in persisted state](https://github.com/digidem/comapeo-mobile/blob/main/src/frontend/hooks/persistedState/usePersistedDraftObservation/index.ts).

We should follow a similar pattern with tracks. Instead of saving [locationHistory to plain zustand state](https://github.com/digidem/comapeo-mobile/pull/271/files#diff-8e3529efd063e458ea9b2c98b258af8cad74a993665ca6390dc59ad823da98bf) we should save it to persisted state.

Also, when the [app open we want to do a check](https://github.com/digidem/comapeo-mobile/blob/dd9101b1765e5d3398b0ab6f5a231a0ba7f049f9/src/frontend/Navigation/AppNavigator.tsx#L83) to see if `locationHistory` exists in persisted state, and if it does, open the SaveTrackScreen. 

^^When we do this  changes the navigation behaviour. When the hit save, they will be [navigated to the map](https://github.com/digidem/comapeo-mobile/blob/8a47fdfaf088fa744f7866aa70c67fdf054e8729/src/frontend/screens/MapScreen/track/saveTrack/SaveTrackHeader.tsx#L60). But the user will now be able to press the back button and go back to the saveTrackScreen. So we need to prevent this by resetting the nav stack on save.

File: src/frontend/screens/MapScreen/track/SaveTrackScreen.tsx
Comment (ErikSin) [inline src/frontend/screens/MapScreen/track/SaveTrackScreen.tsx:1]: We also need to prevent the user from going back by using the back button. When they press the back button it should open the discard track modal

File: src/frontend/screens/MapScreen/track/saveTrack/DiscardTrackModal.tsx
Comment (ErikSin) [inline src/frontend/screens/MapScreen/track/saveTrack/DiscardTrackModal.tsx:1]: We are going to start using this modal for the observations. So can you make the modal `title`, `description` and `discardButton Text` props, so that it can be reused by another component

Comment (bohdanprog) [reply src/frontend/screens/MapScreen/track/saveTrack/DiscardTrackModal.tsx:1]: Certainly, I have completed that task. Please let me know if I can assist with anything else.

File: src/frontend/hooks/tracks/useTracking.ts
Comment (bohdanprog) [reply src/frontend/hooks/tracks/useTracking.ts:44]: fixed

File: src/frontend/screens/MapScreen/track/SaveTrackScreen.tsx
Comment (bohdanprog) [reply src/frontend/screens/MapScreen/track/SaveTrackScreen.tsx:1]: Yeah, good catch, thanks.

File: src/frontend/screens/MapScreen/GPSPermissions/GPSPermissionsEnabled.tsx
Comment (CDFN) [reply src/frontend/screens/MapScreen/GPSPermissions/GPSPermissionsEnabled.tsx:56]: Frequency is about every 3 meters. However if you stay in one place, there are no updates being sent (you can think of it like some kind of debounce). If you travel at least 3 meters you would see there is another location saved. We discussed this behavior on Slack on April 22nd.

File: src/frontend/screens/MapScreen/track/saveTrack/SaveTrackHeader.tsx
Comment (CDFN) [reply src/frontend/screens/MapScreen/track/saveTrack/SaveTrackHeader.tsx:1]: Thanks for adding direct link, indeed it didn't show under previous comment for us. So explanation for this is how the header looks in Figma. We will make it use navigator default header if complying 100% with design isn't that important in this case.

**Comment by achou11** [comment]
> If the user has stopped "tracking" but is in the middle of saving it (aka they are on `SaveTrackScreen`) and they close the app, we want to make sure it opens again on the saveTrackScreen, and that the tracks are persisted in state. In observations acheive this by [saving observations in persisted state](https://github.com/digidem/comapeo-mobile/blob/main/src/frontend/hooks/persistedState/usePersistedDraftObservation/index.ts).
> 
> We should follow a similar pattern with tracks. Instead of saving [locationHistory to plain zustand state](https://github.com/digidem/comapeo-mobile/pull/271/files#diff-8e3529efd063e458ea9b2c98b258af8cad74a993665ca6390dc59ad823da98bf) we should save it to persisted state.
> 
> Also, when the [app open we want to do a check](https://github.com/digidem/comapeo-mobile/blob/dd9101b1765e5d3398b0ab6f5a231a0ba7f049f9/src/frontend/Navigation/AppNavigator.tsx#L83) to see if `locationHistory` exists in persisted state, and if it does, open the SaveTrackScreen.
> 
> ^^When we do this changes the navigation behaviour. When the hit save, they will be [navigated to the map](https://github.com/digidem/comapeo-mobile/blob/8a47fdfaf088fa744f7866aa70c67fdf054e8729/src/frontend/screens/MapScreen/track/saveTrack/SaveTrackHeader.tsx#L60). But the user will now be able to press the back button and go back to the saveTrackScreen. So we need to prevent this by resetting the nav stack on save.

@ErikSin thoughts on this being a follow-up to this PR? The scope of changes required to get the above working might be  large enough to warrant a separate PR to make it easier to review. Not sure if @CDFN @bohdanprog have already started making changes here towards but if they haven't, would put it into consideration

**Comment by ErikSin** [comment]
> @ErikSin thoughts on this being a follow-up to this PR? The scope of changes required to get the above working might be large enough to warrant a separate PR to make it easier to review. Not sure if @CDFN @bohdanprog have already started making changes here towards that but if they haven't, would put it into consideration

Yeah, that sounds good. I do think we should do this sooner, rather than later as it will effect the subsequent Pr, but I think separating it makes sense.

**Comment by CDFN** [comment]
I got some changes but it's fairly easy to move it to another branch. For issue, would be great to summarize it there if you could.

**Comment by CDFN** [comment]
@achou11 does this mean we should merge this one and work on #306 right after? Or just merge #306 into `feat/save-track` and then merge this pr into main?

File: src/frontend/screens/MapScreen/track/saveTrack/SaveTrackHeader.tsx
Comment (achou11) [reply src/frontend/screens/MapScreen/track/saveTrack/SaveTrackHeader.tsx:1]: yeah i think it's preferred to use the existing header for now, that way it'll be easier for us to convert everything at once without gaps of inconsistency

File: src/frontend/screens/MapScreen/GPSPermissions/GPSPermissionsEnabled.tsx
Comment (achou11) [reply src/frontend/screens/MapScreen/GPSPermissions/GPSPermissionsEnabled.tsx:56]: yeah i'm seeing the conversation you're referring to on slack. thanks for pointing that out! seems like i missed it

**Review by achou11 — CHANGES_REQUESTED** [review]
Can this [comment about the navigation header](https://github.com/digidem/comapeo-mobile/pull/271/files/358aa50a66c217596a66c41e3d2f6e31728df326#r1581406217) be addressed before merging?

Otherwise changes seem ready for merging :)

**Comment by bohdanprog** [comment]
> Can this [comment about the navigation header](https://github.com/digidem/comapeo-mobile/pull/271/files/358aa50a66c217596a66c41e3d2f6e31728df326#r1581406217) be addressed before merging?
> 
> Otherwise changes seem ready for merging :)

Hm...but I already changed Header in SaveTrackScreen :D

**Review by achou11 — APPROVED** [review]

**Comment by ErikSin** [comment]
Im going to merge this as I have a PR stacked on this, and I would like to create another PR based on this

**Pull request merged**
